tools/triton-sme-opt/triton-sme-opt.cpp

```diff
@@ -0,0 +1,142 @@
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/InitAllDialects.h"
+#include "mlir/InitAllPasses.h"
+#include "mlir/Interfaces/VectorInterfaces.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassManager.h"
+#include "mlir/Tools/mlir-opt/MlirOptMain.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include <iostream>
+#include "mlir/IR/Operation.h"
+#include <llvm/Support/raw_ostream.h>
+
+
+using namespace mlir;
+
+
+namespace {
+struct OuterProductVectorizationPass : public PassWrapper<OuterProductVectorizationPass, OperationPass<func::FuncOp>> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<vector::VectorDialect, func::FuncDialect>();
+  }
+
+  void runOnOperation() override {
+    func::FuncOp funcOp = getOperation();
+    MLIRContext *context = funcOp.getContext();
+    RewritePatternSet patterns(context);
+
+    // Step 4: Lower vector.multi_reduction to vector.contract (+ some helpful patterns)
+    vector::VectorTransformsOptions vectorTransformsOptions;
+    vectorTransformsOptions.setVectorTransformsOptions(vector::VectorContractLowering::OuterProduct);
+    vector::populateVectorTransferDropUnitDimsPatterns(patterns);
+    vector::populateVectorReductionToContractPatterns(patterns);
+
+    if (failed(applyPatternsAndFoldGreedily(funcOp, std::move(patterns)))) {
+      return signalPassFailure();
+    }
+
+    // Step 5: Lower vector.contract to vector.outerproduct. Also drop unit dims.
+    patterns.clear();
+    vectorTransformsOptions.setVectorTransformsOptions(vector::VectorContractLowering::OuterProduct);
+    vector::populateVectorTransferDropUnitDimsPatterns(patterns);
+
+    if (failed(applyPatternsAndFoldGreedily(funcOp, std::move(patterns)))) {
+      return signalPassFailure();
+    }
+  }
+};
+
+
+  struct MatmulTileConversion: public OpRewritePattern <linalg::MatmulOp> {
+    using OpRewritePattern <linalg::MatmulOp> ::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(linalg::MatmulOp op, PatternRewriter & rewriter) const override {
+
+      SmallVector<int64_t, 3> tileSizes = {4, 4,1}; // Tile sizes for [M, N, K] dimensions tofo
```

Comment:
Hey, I worked on the ArmSME example you linked to (https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/matmul.mlir) - great to see this being adopted in Triton! However, I'm a little confused by this PR. ArmSME requires 2D scalable vectors and tiling, but the vectorization/tiling done here is all fixed-sized, so this won't use any of the ArmSME (or scalable vector) support in MLIR. 

The tile sizes here `4, 4, 1`, which leads to fixed-size vectors (`vector<4x4x1xf32>`), but ArmSME needs 2D scalable sizes. So to target ArmSME you'd need to tile/vectorize (assuming float32) with sizes of `(4 x vscale), (4 x vscale), 1`. This would lead to 2D scalable vectors (`vector<[4]x[4]xf32>`) which then can be lowered to SME tiles (and MOPA instructions). 

Note to target SME you'd also need to make use of the ArmSME dialect lowering passes to convert the 2D-scalable vectors to SME intrinsics (the backend won't do it automatically). 

Also, note that the `matmul.mlir` example is not showing an optimal SME matmul, it's simply testing targeting SME (and uses only a single SME tile, there's actually four SME tiles available for float32).

Reply:
I think I'm about to push a change to address that by making the vectors scalable, I was under the impression that -test-lower-to-arm-sme would auto insert instrinsics since the example you mentioned doesn't explictly say anything in the transform pass except for the -test-lower-to-arm-sme part. I guess I'll have to manually insert ops from the SME dialect but as @banach-space as mentioned, vector dialect is supposed to lower to SME.

Reply:
There are passes to lower from the vector dialect to ArmSME (and those are included in `-test-lower-to-arm-sme`). However, those expect you to tile for SME (which means using scalable vectors and scalable tile sizes), which is not happening in this PR currently, hence my confusion.

Reply:
The syntax in MLIR is not great, but dimension in square brackets it's scalable. So on these lines:
https://github.com/llvm/llvm-project/blob/main/mlir/test/Integration/Dialect/Linalg/CPU/ArmSME/matmul.mlir#L62-L66

The `[4]`s e.g. in `vector_sizes [[4], [4], 1]` means `[4 x vscale, 4 x vscale, 1]`, so those dimensions are scalable. However, what you've done here its equivalent to `vector_sizes [4, 4, 1]`, which means everything is fixed-size (which won't lower to ArmSME).

Reply:
my first approach was just to multiply 4 by cat /proc/sys/abi/sve_default_vector_length
which would give me the systems current SVE length but I think the types themselves need to ScalableVectors. Currently trying to figure out the correct C++ bindings to that

Reply:
Sounds like you are trying to use VLS (Vector Length Specific) vectorisation, but that's not supported. You might be able to short-wire things, but I would recommend using VLA (Vector Length Agnostic) approach instead. So, basically, as Ben suggested - use `[4 x vscale, 4 x vscale, 1]` instead of `[4, 4, 1]`.

Reply:
>  Not sure how to get vscale value in c++

You don't :) That's the whole point of VLA - the actual value of `vscale` is not required. It takes a bit of getting used to, but the whole SVE ISA is designed in a way so that that information is not needed.

In terms of C++ implementation, this could be helpful:
* https://github.com/llvm/llvm-project/blob/982e9022ca4abaad58c693d2b0aba0e908ee2d39/mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp#L2649-L2661

Reply:
Yes, use tilingOptions.setTileSizeComputationFunction() (https://github.com/llvm/llvm-project/blob/982e9022ca4abaad58c693d2b0aba0e908ee2d39/mlir/lib/Dialect/Linalg/TransformOps/LinalgTransformOps.cpp#L2642C1-L2643C81), instead of setTileSizes().

Reply:
this does compile but the vectorization pass fails, this does use vscale, any ideas?


```cpp
  struct MatmulTileConversion: public OpRewritePattern <linalg::MatmulOp> {
    using OpRewritePattern <linalg::MatmulOp> ::OpRewritePattern;

    LogicalResult matchAndRewrite(linalg::MatmulOp op, PatternRewriter & rewriter) const override {
    linalg::LinalgTilingOptions tilingOptions;
    tilingOptions.setTileSizeComputationFunction([&](OpBuilder& b, Operation*) {
      SmallVector<mlir::Value, 4> sizes;
      sizes.reserve(3);
      Location loc = op.getLoc();

      Value vscale = b.create<vector::VectorScaleOp>(loc, b.getIndexType());

      Value tileM = b.create<arith::ConstantIndexOp>(loc, 4);
      Value tileMScaled = b.create<arith::MulIOp>(loc, tileM, vscale);
      sizes.push_back(tileMScaled);

      Value tileN = b.create<arith::ConstantIndexOp>(loc, 4);
      Value tileNScaled = b.create<arith::MulIOp>(loc, tileN, vscale);
      sizes.push_back(tileNScaled);

      Value tileK = b.create<arith::ConstantIndexOp>(loc, 1);
      sizes.push_back(tileK);

      return sizes;
    });

      auto tiledOpResult = tileLinalgOp(rewriter, op, tilingOptions);
      if (failed(tiledOpResult)) {
        std::cout << "TILING FAILED" << std::endl;
        return failure();
      }

      if (failed(linalg::vectorize(rewriter, cast<linalg::LinalgOp> (tiledOpResult->op.getOperation())))) {
        std::cout << "Vectorization FAILED" << std::endl;
        return failure();
      }
      MLIRContext *context = getContext();

      rewriter.replaceOp(op, tiledOpResult->tensorResults);
      return success();

    }
  };
```

Reply:
https://github.com/llvm/llvm-project/blob/d9c855014123a313006adfc873be6f10e997be61/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L1964-L1968

linalg::vectorize() let you specifiy vector sizes and if it's scalable of each dimension. You may want to try passing `ArrayRef<int64_t> inputVectorSizes` and `ArrayRef<bool> inputScalableVecDims`.

Reply:
the pass now works with scalable vectors